src/libcore/ptr.rs

```diff
@@ -10,9 +10,29 @@
 
 // FIXME: talk about offset, copy_memory, copy_nonoverlapping_memory
 
-//! Raw, unsafe pointers, `*const T`, and `*mut T`.
+//! Manually manage memory through raw pointers.
 //!
 //! *[See also the pointer primitive types](../../std/primitive.pointer.html).*
+//!
+//! # Safety
+//!
+//! Many functions in this module take raw pointers as arguments and dereference
+//! them. For this to be safe, these pointers must be valid. A valid pointer
+//! is one that satisfies **all** of the following conditions:
+//!
+//! * The pointer is not null.
+//! * The pointer is not dangling (it does not point to memory which has been
+//!   freed).
+//! * The pointer satisfies [LLVM's pointer aliasing rules].
```

Comment:
This is not very helpful without further explanation on which LLVM aliasing annotations are emitted. That, in turn, may actually change.

I'd expect this to say, at least, that for a pointer to be valid for reading, there must be no live mutable references overlapping with the accessed memory region that this pointer is not based on (as defined in the LLVM rules).  For a pointer to be valid for writing, there must moreover be no live shared references overlapping with the accessed memory region, ignoring memory covered by an `UnsafeCell`.

That may sill not be sufficient, but it's a good start.

Reply:
I also agree with what @rkruppe wrote, this should contain an explicit note that these rules are still work-in-progress and hence subject to change and if there is any doubt, people should assume that what they are doing is probably UB.

Reply:
Which sections do you plan to link to?

Reply:
Here's a list of possible candidates:
* [Dereferencing Raw Pointers in book](https://doc.rust-lang.org/book/second-edition/ch19-01-unsafe-rust.html#dereferencing-a-raw-pointer)
* [Undefined Behavior in reference](https://doc.rust-lang.org/reference/behavior-considered-undefined.html)
* [Aliasing in nomicon](https://doc.rust-lang.org/nomicon/aliasing.html)
* [Uninitialized Memory in nomicon](https://doc.rust-lang.org/nomicon/unchecked-uninit.html)
* [LLVM Pointer Aliasing](https://llvm.org/docs/LangRef.html#pointer-aliasing-rules): cited by the reference, and the only real resource for 

I propose linking to the first two. The first one contains some useful general info, and the second one references (`&mut`) aliasing, uninitialized memory and (C-style) pointer aliasing.

I'd like to defer further definitions to a later PR (this one is already suffering from scope creep), which would add a nomicon page discussing the validity of pointer accesses:

* In a vacuum, pointer validity depends on three factors, the value of the pointer, the size of the access, the type of the access (read or write).
* Additionally, a pointer is only valid for an access of a given size if it does not alias with a `&mut` reference (not entirely sure about the subtleties of this one).
* Reading or writing to an invalid pointer is undefined behavior
* A pointer is valid for reads if it points to uninitialized memory, but no assumptions can be made about the resulting value, or the results of any operations on the resulting value. Notably one cannot assume that it remains the same across multiple uses. For example (`x ^ x == 0`) may not be `true`, and `x & 0 == 0` may not be `true` if`x` was the result of a read of uninitialized memory. ( believe that this is stricter than LLVM's `undef`--for which the result of the first is not defined, but the result of the second is defined--but more relaxed than the current reference, which states that any read of uninitialized memory is UB. This may require a proper PR.
* The result of the casts `&x as *const _` and `&mut x as *mut _` is always a valid pointer (This is only true if creating a reference from an invalid pointer is already undefined behavior).
* Some stuff about pointer arithmetic and memory mapped I/O, which I am not currently ready to think about :)

Reply:
I'd add a reference to the aliasing page in the nomicon, but I believe you have such a reference included currently?

I see uninitialized memory as being orthogonal to what you are discussing here -- "uninitialized" is just yet another "value" that can be read (IOW, a byte is either a sequence of 8 bits, or "uninitialized"). This agrees with your characterization. It means that in principle, unintiailized memory does not have to be mentioned at all in the pointer validity discussion -- though it's probably worth clarifying this point as this question will come up.